File: baselibrary/src/main/java/com/sscl/baselibrary/widget/GifView.java
package com.sscl.baselibrary.widget;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Movie;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.sscl.baselibrary.R;


/**
 * 自定义用于显示gif动画的控件
 *
 * @author jackie
 */
public class GifView extends View {

    /**
     * 默认为1秒
     */
    private static final int DEFAULT_MOVIE_DURATION = 1000;

    private int mMovieResourceId;

    private Movie mMovie;

    private long mMovieStart;

    private int mCurrentAnimationTime = 0;

    private float mLeft;

    private float mTop;

    private float mScale;

    private int mMeasuredMovieWidth;

    private int mMeasuredMovieHeight;

    private boolean mVisible = true;

    private volatile boolean mPaused = false;

    /**
     * 构造方法
     *
     * @param context 上下文
     */
    public GifView(@NonNull Context context) {
        this(context, null);
    }

    /**
     * 构造方法
     *
     * @param context 上下文
     * @param attrs   参数
     */
    public GifView(@NonNull Context context,@Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    /**
     * 构造方法
     *
     * @param context  上下文
     * @param attrs    参数
     * @param defStyle 默认风格
     */
    public GifView(@NonNull Context context,@Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        setViewAttributes(context, attrs, defStyle);
    }

    @SuppressLint("NewApi")
    private void setViewAttributes(@NonNull Context context,@Nullable AttributeSet attrs,
                                   int defStyle) {
        setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        // 从描述文件中读出gif的值，创建出Movie实例
        final TypedArray array = context.obtainStyledAttributes(attrs,
                R.styleable.GifView, defStyle, R.style.Widget_GifView);
        mMovieResourceId = array.getResourceId(R.styleable.GifView_gif, -1);
        mPaused = array.getBoolean(R.styleable.GifView_paused, false);
        array.recycle();
        if (mMovieResourceId != -1) {
            mMovie = Movie.decodeStream(getResources().openRawResource(
                    mMovieResourceId));
        }
    }

    /**
     * 设置gif图资源
     *
     * @param movieResId 要设置的gif图资源
     */
    @SuppressWarnings("unused")
    public void setMovieResource(int movieResId) {
        this.mMovieResourceId = movieResId;
        mMovie = Movie.decodeStream(getResources().openRawResource(
                mMovieResourceId));
        requestLayout();
    }

    /**
     * 设置gif的动画(视频)
     *
     * @param movie 要设置的gif动画(视频)
     */
    @SuppressWarnings("unused")
    public void setMovie(Movie movie) {
        this.mMovie = movie;
        requestLayout();
    }

    /**
     * 获取gif的动画(视频)
     *
     * @return gif的动画(视频)
     */
    @Nullable
    @SuppressWarnings("unused")
    public Movie getMovie() {
        return mMovie;
    }

    /**
     * 设置gif的动画(视频)的时间
     *
     * @param time 要设置的gif动画(视频)时间
     */
    @SuppressWarnings("unused")
    public void setMovieTime(int time) {
        mCurrentAnimationTime = time;
        invalidate();
    }

    /**
     * 设置暂停
     *
     * @param paused 要设置的暂停标识(true表示暂停，false表示播放)
     */
    public void setPaused(boolean paused) {
        this.mPaused = paused;
        if (!paused) {
            mMovieStart = android.os.SystemClock.uptimeMillis()
                    - mCurrentAnimationTime;
        }
        invalidate();
    }

    /**
     * 判断gif图是否停止了
     *
     * @return true表示已经停止了，false表示没有停止
     */
    @SuppressWarnings("unused")
    public boolean isPaused() {
        return this.mPaused;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (mMovie != null) {
            int movieWidth = mMovie.width();
            int movieHeight = mMovie.height();
            int maximumWidth = MeasureSpec.getSize(widthMeasureSpec);
            float scaleWidth = (float) movieWidth / (float) maximumWidth;
            mScale = 1f / scaleWidth;
            mMeasuredMovieWidth = maximumWidth;
            mMeasuredMovieHeight = (int) (movieHeight * mScale);
            setMeasuredDimension(mMeasuredMovieWidth, mMeasuredMovieHeight);
        } else {
            setMeasuredDimension(getSuggestedMinimumWidth(),
                    getSuggestedMinimumHeight());
        }
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        super.onLayout(changed, l, t, r, b);
        mLeft = (getWidth() - mMeasuredMovieWidth) / 2f;
        mTop = (getHeight() - mMeasuredMovieHeight) / 2f;
        mVisible = getVisibility() == View.VISIBLE;
    }

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        if (mMovie != null) {
            if (!mPaused) {
                updateAnimationTime();
                drawMovieFrame(canvas);
                invalidateView();
            } else {
                drawMovieFrame(canvas);
            }
        }
    }

    private void invalidateView() {
        if (mVisible) {
            postInvalidateOnAnimation();
        }
    }

    private void updateAnimationTime() {
        long now = android.os.SystemClock.uptimeMillis();
        // 如果第一帧，记录起始时间
        if (mMovieStart == 0) {
            mMovieStart = now;
        }
        // 取出动画的时长
        int dur = mMovie.duration();
        if (dur == 0) {
            dur = DEFAULT_MOVIE_DURATION;
        }
        // 算出需要显示第几帧
        mCurrentAnimationTime = (int) ((now - mMovieStart) % dur);
    }

    private void drawMovieFrame(Canvas canvas) {
        // 设置要显示的帧，绘制即可
        mMovie.setTime(mCurrentAnimationTime);
//        canvas.save(Canvas.MATRIX_SAVE_FLAG);
        canvas.save();
        canvas.scale(mScale, mScale);
        mMovie.draw(canvas, mLeft / mScale, mTop / mScale);
        canvas.restore();
    }

    @SuppressLint("NewApi")
    @Override
    public void onScreenStateChanged(int screenState) {
        super.onScreenStateChanged(screenState);
        mVisible = screenState == SCREEN_STATE_ON;
        invalidateView();
    }

    @SuppressLint("NewApi")
    @Override
    protected void onVisibilityChanged(@NonNull View changedView, int visibility) {
        super.onVisibilityChanged(changedView, visibility);
        mVisible = visibility == View.VISIBLE;
        invalidateView();
    }

    @Override
    protected void onWindowVisibilityChanged(int visibility) {
        super.onWindowVisibilityChanged(visibility);
        mVisible = visibility == View.VISIBLE;
        invalidateView();
    }

}

File: baselibrary/src/main/java/com/sscl/baselibrary/widget/GifView.kt
package com.sscl.baselibrary.widget

import kotlin.jvm.JvmOverloads
import com.sscl.baselibrary.R
import android.annotation.SuppressLint
import android.content.res.TypedArray
import kotlin.jvm.Volatile
import android.content.*
import android.graphics.*
import android.os.*
import android.util.AttributeSet
import android.view.*

/**
 * 自定义用于显示gif动画的控件
 *
 * @author jackie
 */
class GifView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : View(context, attrs, defStyle) {
    private var mMovieResourceId: Int = 0
    private var mMovie: Movie? = null
    private var mMovieStart: Long = 0
    private var mCurrentAnimationTime: Int = 0
    private var mLeft: Float = 0f
    private var mTop: Float = 0f
    private var mScale: Float = 0f
    private var mMeasuredMovieWidth: Int = 0
    private var mMeasuredMovieHeight: Int = 0
    private var mVisible: Boolean = true

    @Volatile
    private var mPaused: Boolean = false
    @SuppressLint("NewApi")
    private fun setViewAttributes(
        context: Context, attrs: AttributeSet?,
        defStyle: Int
    ) {
        setLayerType(LAYER_TYPE_SOFTWARE, null)
        // 从描述文件中读出gif的值，创建出Movie实例
        val array: TypedArray = context.obtainStyledAttributes(
            attrs,
            R.styleable.GifView, defStyle, R.style.Widget_GifView
        )
        mMovieResourceId = array.getResourceId(R.styleable.GifView_gif, -1)
        mPaused = array.getBoolean(R.styleable.GifView_paused, false)
        array.recycle()
        if (mMovieResourceId != -1) {
            mMovie = Movie.decodeStream(
                getResources().openRawResource(
                    mMovieResourceId
                )
            )
        }
    }

    /**
     * 设置gif图资源
     *
     * @param movieResId 要设置的gif图资源
     */
    fun setMovieResource(movieResId: Int) {
        mMovieResourceId = movieResId
        mMovie = Movie.decodeStream(
            getResources().openRawResource(
                mMovieResourceId
            )
        )
        requestLayout()
    }
    /**
     * 获取gif的动画(视频)
     *
     * @return gif的动画(视频)
     */
    /**
     * 设置gif的动画(视频)
     *
     * @param movie 要设置的gif动画(视频)
     */
    var movie: Movie?
        get() {
            return mMovie
        }
        set(movie) {
            mMovie = movie
            requestLayout()
        }

    /**
     * 设置gif的动画(视频)的时间
     *
     * @param time 要设置的gif动画(视频)时间
     */
    fun setMovieTime(time: Int) {
        mCurrentAnimationTime = time
        invalidate()
    }
    /**
     * 判断gif图是否停止了
     *
     * @return true表示已经停止了，false表示没有停止
     */
    /**
     * 设置暂停
     *
     * @param paused 要设置的暂停标识(true表示暂停，false表示播放)
     */
    var isPaused: Boolean
        get() {
            return mPaused
        }
        set(paused) {
            mPaused = paused
            if (!paused) {
                mMovieStart = (SystemClock.uptimeMillis()
                        - mCurrentAnimationTime)
            }
            invalidate()
        }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (mMovie != null) {
            val movieWidth: Int = mMovie!!.width()
            val movieHeight: Int = mMovie!!.height()
            val maximumWidth: Int = MeasureSpec.getSize(widthMeasureSpec)
            val scaleWidth: Float = movieWidth.toFloat() / maximumWidth.toFloat()
            mScale = 1f / scaleWidth
            mMeasuredMovieWidth = maximumWidth
            mMeasuredMovieHeight = (movieHeight * mScale).toInt()
            setMeasuredDimension(mMeasuredMovieWidth, mMeasuredMovieHeight)
        } else {
            setMeasuredDimension(
                getSuggestedMinimumWidth(),
                getSuggestedMinimumHeight()
            )
        }
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        mLeft = (getWidth() - mMeasuredMovieWidth) / 2f
        mTop = (getHeight() - mMeasuredMovieHeight) / 2f
        mVisible = getVisibility() == VISIBLE
    }

    override fun onDraw(canvas: Canvas) {
        if (mMovie != null) {
            if (!mPaused) {
                updateAnimationTime()
                drawMovieFrame(canvas)
                invalidateView()
            } else {
                drawMovieFrame(canvas)
            }
        }
    }

    private fun invalidateView() {
        if (mVisible) {
            postInvalidateOnAnimation()
        }
    }

    private fun updateAnimationTime() {
        val now: Long = SystemClock.uptimeMillis()
        // 如果第一帧，记录起始时间
        if (mMovieStart == 0L) {
            mMovieStart = now
        }
        // 取出动画的时长
        var dur: Int = mMovie!!.duration()
        if (dur == 0) {
            dur = DEFAULT_MOVIE_DURATION
        }
        // 算出需要显示第几帧
        mCurrentAnimationTime = ((now - mMovieStart) % dur).toInt()
    }

    private fun drawMovieFrame(canvas: Canvas) {
        // 设置要显示的帧，绘制即可
        mMovie!!.setTime(mCurrentAnimationTime)
        //        canvas.save(Canvas.MATRIX_SAVE_FLAG);
        canvas.save()
        canvas.scale(mScale, mScale)
        mMovie!!.draw(canvas, mLeft / mScale, mTop / mScale)
        canvas.restore()
    }

    @SuppressLint("NewApi")
    public override fun onScreenStateChanged(screenState: Int) {
        super.onScreenStateChanged(screenState)
        mVisible = screenState == SCREEN_STATE_ON
        invalidateView()
    }

    @SuppressLint("NewApi")
    override fun onVisibilityChanged(changedView: View, visibility: Int) {
        super.onVisibilityChanged(changedView, visibility)
        mVisible = visibility == VISIBLE
        invalidateView()
    }

    override fun onWindowVisibilityChanged(visibility: Int) {
        super.onWindowVisibilityChanged(visibility)
        mVisible = visibility == VISIBLE
        invalidateView()
    }

    companion object {
        /**
         * 默认为1秒
         */
        private val DEFAULT_MOVIE_DURATION: Int = 1000
    }
    /**
     * 构造方法
     *
     * @param context  上下文
     * @param attrs    参数
     * @param defStyle 默认风格
     */
    /**
     * 构造方法
     *
     * @param context 上下文
     * @param attrs   参数
     */
    /**
     * 构造方法
     *
     * @param context 上下文
     */
    init {
        setViewAttributes(context, attrs, defStyle)
    }
}